### ISSUE TITLE
mptt 0.7.2 forces saving parent before creating child

### ISSUE THREAD

**Issue by erez-o** [django-mptt/django-mptt#373]
In django-mptt 0.6.1, django 1.7, creating a child, linking it to a parent, without the parent being saved was allowed:

For example, the following slightly different version than the [docs](http://django-mptt.github.io/django-mptt/tutorial.html#getting-started) was allowed:

```
from myapp.models import Genre
rock = Genre(parent=None, name="Rock")
hard_rock = Genre(parent = rock, name="HardRock")
```

However, doing the same in django-mptt 0.7.2, django 1.8 results in the following Traceback :
```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/erez/Envs/env1/local/lib/python2.7/site-packages/mptt/models.py", line 378, in __init__
    super(MPTTModel, self).__init__(*args, **kwargs)
  File "/home/erez/Envs/env1/local/lib/python2.7/site-packages/django/db/models/base.py", line 468, in __init__
    setattr(self, field.name, rel_obj)
  File "/home/erez/Envs/env1/local/lib/python2.7/site-packages/django/db/models/fields/related.py", line 668, in __set__
    (value, self.field.rel.to._meta.object_name)
ValueError: Cannot assign "<Genre: Genre object>": "Genre" instance isn't saved in the database.
```

Only If I alter the above example, and save the the parent before linking the child, it works without throwing an exception:
```
from myapp.models import Genre
rock = Genre(parent=None, name="Rock")
rock.save()
hard_rock = Genre(parent = rock, name="HardRock")
```

Delaying save is an extremely important feature of mptt, especially if the tree size is in the hundreds or thousands.

Any help is appreciated :)

**Comment by matthiask** [comment]
That's actually not django-mptt's doing, but Django's:

https://docs.djangoproject.com/en/1.8/ref/models/fields/#django.db.models.ForeignKey.allow_unsaved_instance_assignment

if you need this feature you should set the flag `allow_unsaved_instance_assignment` yourself.

**Comment by erez-o** [comment]
Thx, great catch, fixed.

While it solved my problem, I don't think its a good solution for mptt that most users will create their own.
```
class UnsavedTreeForeignKey(TreeForeignKey):
    allow_unsaved_instance_assignment = True
```

As I see it, MPTT was designed for big deep trees where traversing a tree costs a lot of computation. No need for it if the tree is small or shallow.
Repeatedly saving during creation is not advised, hence the creation of `delay_mptt_updates` and `disable_mptt_updates`, which now will cause problems for cases like in this [SO question](http://stackoverflow.com/questions/12661488/how-optimize-adding-new-nodes-in-django-mptt)

My 2cents - Either: 
1) Make `allow_unsaved_instance_assignment=True` the default for `TreeManager`
2) Create mptt.fields.UnsavedTreeForeignKey
3) Change the docs. 

Hope this feedback helps :)

**Comment by craigds** [comment]
I don't really see the difference. You're going to have to save each node once regardless, right? What's the difference between:

    a.save()
    b = Genre(parent=a)
    b.save()

and

    b = Genre(parent=a)
    a.save()
    b.save()
    
?

In both cases you're saving each node once. And you *have* to save `a` before you save `b`, otherwise you won't have a primary-key value for it to save in `b`'s parent field.

**Comment by craigds** [comment]
agree with @matthiask

**Comment by erez-o** [comment]
@craigds, maybe my specific case is clouding me and doesn't represent the norm for other users of mptt.

In my case, I'm reading an xml file, and parsing only relevant info into a tree. It's much faster for me to create all leaves bit by bit than to create the logic for each leaf separately. Since save() is a performance bottleneck, I save at the end.

In the linked SO question, you gave an example of how to increase the performance of bulk saving/updating:
```
with transaction.atomic():
    with MyMPTTModel.objects.disable_mptt_updates():
        for new_record in new_records:
            new_node = MyMPTTModel(..., parent=get_parent(new_record))
            new_node.save()
    MyMPTTModel.objects.rebuild()
```

Basically, the knowledge of how to create each leaf needs to be inside these two with statements. Any mptt user that can live with that won't need `allow_unsaved_instance_assignment = True`

In my case, doing it is very suboptimal. 
I'm not sure now that it represents the norm, you are a better judge than I am.
